mlflow/transformers/signature.py

```diff
@@ -0,0 +1,202 @@
+    ),
+    transformers.AutomaticSpeechRecognitionPipeline: ModelSignature(
+        inputs=Schema([ColSpec(DataType.binary)]),
+        outputs=Schema([ColSpec(DataType.string)]),
+    ),
+    transformers.AudioClassificationPipeline: ModelSignature(
+        inputs=Schema([ColSpec(DataType.binary)]),
+        outputs=Schema(
+            [ColSpec(DataType.double, name="score"), ColSpec(DataType.string, name="label")]
+        ),
+    ),
+    transformers.TableQuestionAnsweringPipeline: ModelSignature(
+        inputs=Schema(
+            [ColSpec(DataType.string, name="query"), ColSpec(DataType.string, name="table")]
+        ),
+        outputs=Schema([ColSpec(DataType.string)]),
+    ),
+    transformers.QuestionAnsweringPipeline: ModelSignature(
+        inputs=Schema(
+            [ColSpec(DataType.string, name="question"), ColSpec(DataType.string, name="context")]
+        ),
+        outputs=Schema([ColSpec(DataType.string)]),
+    ),
+    transformers.FeatureExtractionPipeline: ModelSignature(
+        inputs=Schema([ColSpec(DataType.string)]),
+        outputs=Schema([TensorSpec(np.dtype("float64"), [-1], "double")]),
+    ),
+}
+
+
+def infer_or_get_default_signature(
+    pipeline, example=None, model_config=None, flavor_config=None
+) -> ModelSignature:
+    """
+    Assigns a default ModelSignature for a given Pipeline type that has pyfunc support. These
+    default signatures should only be generated and assigned when saving a model iff the user
+    has not supplied a signature.
+    For signature inference in some Pipelines that support complex input types, an input example
+    is needed.
+    """
+    if example:
+        try:
+            timeout = MLFLOW_INPUT_EXAMPLE_INFERENCE_TIMEOUT.get()
+            if timeout and not _IS_UNIX:
+                timeout = None
+                _logger.warning(
+                    "On Windows, timeout is not supported for model signature inference. "
+                    "Therefore, the operation is not bound by a timeout and may hang indefinitely. "
+                    "If it hangs, please consider specifying the signature manually."
+                )
+            return _infer_signature_with_prediction(
+                pipeline, example, model_config, flavor_config, timeout
+            )
+        except MlflowTimeoutError:
+            _logger.warning(
+                "Attempted to generate a signature for the saved model but prediction operation "
+                f"timed out after {timeout} seconds. Falling back to the default signature for the "
+                "pipeline. You can specify a signature manually or increase the timeout "
+                f"by setting the environment variable {MLFLOW_INPUT_EXAMPLE_INFERENCE_TIMEOUT}"
+            )
+        except Exception as e:
+            _logger.error(
+                "Attempted to generate a signature for the saved model or pipeline "
+                f"but encountered an error: {e}"
+            )
+            raise
```

Comment:
Should we raise the exception or just return None?

Reply:
I think we should raise, cuz this case highly likely be an critical issue of model prediction that would cause same issue after loaded/served?

Reply:
Good point. Can we attempt to use the fallback in the case of any failure that occurs and only raise if a signature cannot be generated at all? 
- remove the raise
- modify the warning in line 125 to raise an MLflowException

The reason being is that if a signature is not generated for these models on Databricks, they won't be eligible for registration in UC and won't be able to be submitted to model serving.

Reply:
I see, but should we allow all errors during signature inference? For example, we raise MlflowException when the given model is not Pipeline instance (L145). Also whatever happens in this prediction will happen in production after serving I guess. Solving an issue in model serving is kinda hard

```
_TransformersWrapper(
        pipeline=pipeline, model_config=model_config, flavor_config=flavor_config
    ).predict(data, params=params)
```

What about block for those exceptions at least to fail first, while allowing fall back for any errors from our code i.e. signature inference logic?

Reply:
I think if there is any errors during prediction result generation, the input example might be wrong (or the model has some problem), while the signature doesn't necessarily requires an output schema.

Reply:
> the input example might be wrong (or the model has some problem)

Yeah this is what I'm worried about, it's better telling users that "hey sth is wrong with your model or example". But I agree that the signature itself doesn't necessary need the output, so probably such validation is beyond the responsibility of this function. Will update to fallback not throwing (which I realize that same as what we do for requirement inference as well).

Reply:
Reverted change for the case where no default fallback signature is found. If we change it from warning to exception, it prevent customers to save custom pipeline and caused failure for test case `test_invalid_task_inference_raises_error`. While it might be not ideal for UC experience, I keep the original behavior (just warn and return no signature) in the scope of this PR. I can do follow-up if necessary.